### PR TITLE
Added ability to adding arbitrary vectors to the residual stream 

### PR DESCRIPTION
**Pull request by levmckinney**

### CODE REVIEW

**Review by rhaps0dy — COMMENTED** [review]
EDIT: what's the difference between this and #112 ??

File: tuned_lens/model_surgery.py [hunk 1]
```diff
@@ -211,8 +211,8 @@ def permute_layers_(model: th.nn.Module, indices: list[int]):
 
 @contextmanager
 def replace_layers(
-    model: T, indices: list[int], replacements: list[th.nn.Module]
-) -> Generator[T, None, None]:
+    model: Model, indices: list[int], replacements: list[th.nn.Module]
+) -> Generator[None, None, None]:
```
Comment (rhaps0dy) [inline tuned_lens/model_surgery.py:215]: The model is modified in place right? That's why returning it is not needed.

Comment (rhaps0dy) [reply tuned_lens/model_surgery.py:215]: Why do other functions (e.g. `delete_layers`, `permute_layers`, `assign_key_path`) still yield the modified model?

File: tuned_lens/model_surgery.py [hunk 2]
```diff
@@ -221,6 +221,87 @@ def replace_layers(
...
+    def __init__(
+        self,
+        base: th.nn.Module,
+        vector: th.Tensor,
+        get_hidden: Callable[[Any], th.Tensor],
+        set_hidden: Callable[[Any, th.Tensor], Any],
+    ):
+        super().__init__()
+        self.base = base
+        self.vector = nn.Parameter(vector.clone())
```
Comment (rhaps0dy) [inline tuned_lens/model_surgery.py:239]: Is this intended to allow optimizers to optimize the steering vector?
- If yes, how can optimizers list the parameters? (upon reading more of the code, it's possible by calling `model.parameters()` inside the context.
- if not, this should be `register_buffer()` or have the arg `requires_grad=False`

File: tuned_lens/model_surgery.py [hunk 3]
```diff
@@ -221,6 +221,87 @@ def replace_layers(
...
+        hidden = self.get_hidden(outputs)
+        hidden = hidden + self.vector
+        outputs = self.set_hidden(outputs, hidden)
+        return outputs
+
+
+@contextmanager
+def add_steering_vector(
+    model: Model,
+    indices: list[int],
```
Comment (rhaps0dy) [inline tuned_lens/model_surgery.py:254]: What happens if indices are repeated? As I understand it, the last value assigned to an index is picked (due to how `replace_layers` is written), but this should be asserted out or documented.

File: tuned_lens/model_surgery.py [hunk 4]
```diff
@@ -221,6 +221,87 @@ def replace_layers(
...
+    steering_vectors: Union[list[th.Tensor], th.Tensor],
+) -> Generator[None, None, None]:
+    """In the context modify the model by adding a steering vectors to the given layers.
+
+    Args:
+        model: The model to apply the steering vector to.
+        indices: The indices of the layers to modify.
+        steering_vectors: The steering vector to add to each layer.
```
Comment (rhaps0dy) [inline tuned_lens/model_surgery.py:262]:
```suggestion
        steering_vectors: The steering vector to add to each layer. If it is a single
            Tensor, it gets added to all the layers with indices in `indices`. 
```

File: tuned_lens/nn/lenses.py [hunk 5]
```diff
@@ -316,3 +316,66 @@ def forward(self, h: th.Tensor, idx: int) -> th.Tensor:
     def __len__(self) -> int:
         """Return the number of layer translators in the lens."""
         return len(self.layer_translators)
+
+    @th.inference_mode()
```
Comment (rhaps0dy) [inline tuned_lens/nn/lenses.py:320]: Why isn't this part already in `main`? Maybe you should merge main into this branch before requesting review again.

**Review by rhaps0dy — CHANGES_REQUESTED** [review]
Please at least merge `main` into this. Otherwise I'd appreciate replies / modifications.

File: tuned_lens/model_surgery.py (hunk 1 above)
Comment (levmckinney) [reply tuned_lens/model_surgery.py:215]: Ah good point I'll change those as well!

File: tuned_lens/model_surgery.py (hunk 2 above)
Comment (levmckinney) [reply tuned_lens/model_surgery.py:239]: Okay I will change this

File: tuned_lens/model_surgery.py (hunk 3 above)
Comment (norabelrose) [reply tuned_lens/model_surgery.py:254]: Yeah we should probably add a `assert len(set(indices)) == len(indices)` check or something

File: tuned_lens/model_surgery.py (hunk 1 above)
Comment (norabelrose) [reply tuned_lens/model_surgery.py:215]: FWIW the reason I initially made it return the model was to mirror PyTorch in-place functions and methods which, for better or worse, also return the tensor. But I don't actually care as long as we're consistent.

**Comment by levmckinney** [comment]
I'm not supper happy with the interface for this and I don't think this is critical for 1.0. Thus I'm closing this PR.

**Pull request closed without merging**